app/controllers/declaration/DeclarationChoiceController.scala

```diff
@@ -0,0 +1,105 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.declaration
+
+import java.time.Instant
+
+import connectors.exchange.ExportsDeclarationExchange
+import controllers.actions.AuthAction
+import forms.declaration.DeclarationChoice
+import forms.declaration.DeclarationChoice._
+import javax.inject.Inject
+import models.DeclarationType.DeclarationType
+import models.requests.ExportsSessionKeys
+import models.{DeclarationStatus, Mode}
+import play.api.Logger
+import play.api.data.Form
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import services.cache.ExportsCacheService
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.controller.FrontendController
+import views.html.declaration.declaration_choice
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class DeclarationChoiceController @Inject()(
+  authenticate: AuthAction,
+  override val exportsCacheService: ExportsCacheService,
+  mcc: MessagesControllerComponents,
+  choicePage: declaration_choice
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with I18nSupport with ModelCacheable {
+
+  private val logger = Logger(this.getClass)
+
+  def displayPage: Action[AnyContent] = authenticate.async { implicit request =>
+    request.declarationId match {
+      case Some(id) =>
+        exportsCacheService.get(id).map(_.map(_.`type`)).map {
+          case Some(data) => Ok(choicePage(form().fill(DeclarationChoice(data))))
+          case _          => Ok(choicePage(form()))
+        }
+      case _ => Future.successful(Ok(choicePage(form())))
+    }
+
+  }
+
+  def submitChoice(): Action[AnyContent] = authenticate.async { implicit request =>
+    form()
+      .bindFromRequest()
+      .fold(
+        (formWithErrors: Form[DeclarationChoice]) => Future.successful(BadRequest(choicePage(formWithErrors))),
+        choice => {
+          val declarationType = choice.value
+
+          request.declarationId match {
+            case Some(id) =>
+              updateDeclarationType(id, declarationType).map { _ =>
+                Redirect(controllers.declaration.routes.DispatchLocationController.displayPage(Mode.Normal))
+              }
+            case _ =>
+              create(declarationType) map { created =>
+                Redirect(controllers.declaration.routes.DispatchLocationController.displayPage(Mode.Normal))
+                  .addingToSession(ExportsSessionKeys.declarationId -> created.id)
+              }
+          }
+        }
+      )
+  }
+
+  private def updateDeclarationType(id: String, `type`: DeclarationType)(implicit hc: HeaderCarrier) =
+    exportsCacheService.get(id).map(_.map(_.copy(`type` = `type`))).flatMap {
```

Comment:
it would be nice to move `_.copy(`type` = `type`)` as named method on `ExportsDeclaration`

Reply:
Yes - it is a bit of a mouthful - I just moved it from the ChoiceController.

Reply:
Done - not sure it was worth it but it was low effort.